runtime/reachable.c

```diff
@@ -0,0 +1,413 @@
+      rv = "fseek failure";
+      break;
+    }
+    len = ftell(exe_file);
+    if (len == -1) {
+      rv = "ftell failure";
+      break;
+    }
+    if (fseek(exe_file, 0L, SEEK_SET) != 0) {
+      rv = "2nd fseek failure";
+      break;
+    }
+    buf = malloc(len);
+    if (buf == NULL) {
+      rv = "malloc failure";
+      break;
+    }
+    elf_info->buf = buf;
+    read_len = fread(buf, sizeof(unsigned char), len, exe_file);
+    if (read_len != len) {
+      rv = "fread failure";
+      break;
+    }
+
+    fh = (void *)buf;
+    if (memcmp(fh->e_ident, ELF, sizeof(ELF)) != 0) {
+      rv = "not an ELF file";
+      break;
+    }
+
+    is32Bit = fh->e_ident[EI_CLASS] == 1;
+    if (is32Bit ^ (sizeof(ELF_Uint) == 4)) {
+      rv = "32/64-bit mismatch in executable file";
+      break;
+    }
+
+    p = (char *)&my_endian;
+    *p = 1;
+    isLittleEndian = fh->e_ident[EI_DATA] == 1;
+    if (isLittleEndian ^ (my_endian == 1)) {
+      rv = "endianness mismatch in executable file";
+      break;
+    }
+
+    if (fh->e_ident[EI_VERSION] != 1 || fh->e_version != 1) {
+      rv = "not ELF V1";
+      break;
+    }
+
+    section_hdrs = (void *)buf + fh->e_shoff;
+    shrstrtab_hdr = &section_hdrs[fh->e_shstrndx];
+    elf_info->shrstrtab = buf + shrstrtab_hdr->sh_offset;
+
+    sh = find_section(elf_info, ".strtab");
+    if (sh == NULL) {
+      rv = "Missing .strtab section";
+      break;
+    }
+    elf_info->strtab = buf + sh->sh_offset;
+    
+    sh = find_section(elf_info, ".symtab");
+    if (sh == NULL) {
+      rv = "Missing .symtab section";
+      break;
+    }
+    elf_info->symtab = (Elf_Sym *)(buf + sh->sh_offset);
+    symtab_entsize = sh->sh_entsize;
+    elf_info->symtab_num_ents = sh->sh_size / symtab_entsize;
+
+    if (symtab_entsize != sizeof (Elf_Sym)) {
+      rv = "Symtab entry size mismatch";
+      break;
+    }
+
+    reloc_sym = caml_reloc_sym;  // avoid "unused variable" warning
+    reloc_sym = caml_get_elf_symbol_by_name(elf_info, "caml_reloc_sym");
+    if (reloc_sym == -1) {
+      rv = "Can't find 'caml_reloc_sym' symbol";
+      break;
+    }
+    elf_info->relocation = ((ELF_Uint)&caml_reloc_sym) - reloc_sym;
+  } while (0);
+
+  if (exe_file != NULL)
+    fclose(exe_file);
+  if (strcmp(rv, "") != 0 && buf != NULL) {
+    elf_info->buf = NULL;
+    free(buf);
+  }
+  return rv;
+}
+
+char *caml_get_elf_symbol_by_address(Elf_info *elf_info, value *addr) {
+  ELF_Uint sym_unrelocated;
+  Elf_Sym *sym_ent;
+  
+  if (elf_info->buf == NULL) {
+    printf("call elf_info_setup before calling caml_get_elf_symbol_by_address\n");
+    return "???";
+  }
+
+  sym_unrelocated = (ELF_Uint)addr - elf_info->relocation;
+  for (int i = 0; i < elf_info->symtab_num_ents; i++) {
+    sym_ent = &elf_info->symtab[i];
+    if (sym_ent->st_value == sym_unrelocated)
+      return strdup(elf_info->strtab + sym_ent->st_name);
+  }
+
+  return "???";  // not found
+}
+
+static struct {
```

Comment:
These global variables accessed by the callback below make `Gc.print_reachable` not thread safe, meaning having 2 calls active at the same time on different threads won't work.  This seems an unlikely use case.  It made sense to call some of the existing routines with my callbacks rather than duplicate that code.  If necessary I could document this or add locking or thread variables.